File: python/exercises/practice/unit_test_c2d4ebc9_9926_4930_a672_6c82fbd231e8/tx_orchestrator/.meta/transaction_coordinator.py
import threading
import concurrent.futures
import time
from typing import List, Callable

class TransactionCoordinator:
    def __init__(self):
        self.lock = threading.Lock()
        self.active_transactions = {}

    def execute_transaction(self, services: List[object], tx_id: str, timeout: int) -> bool:
        with self.lock:
            if tx_id in self.active_transactions:
                raise ValueError(f"Transaction {tx_id} already exists")
            self.active_transactions[tx_id] = {'status': 'pending', 'services': services}

        try:
            # Phase 1: Prepare
            prepared = self._prepare_phase(services, tx_id, timeout)
            if not prepared:
                self._rollback_phase(services, tx_id, timeout)
                return False

            # Phase 2: Commit
            committed = self._commit_phase(services, tx_id, timeout)
            if not committed:
                self._rollback_phase(services, tx_id, timeout)
                return False

            return True
        finally:
            with self.lock:
                self.active_transactions.pop(tx_id, None)

    def _prepare_phase(self, services: List[object], tx_id: str, timeout: int) -> bool:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = []
            for service in services:
                futures.append(executor.submit(self._call_service_operation, 
                                             service.prepare, tx_id, timeout))

            results = []
            for future in concurrent.futures.as_completed(futures):
                try:
                    result = future.result()
                    results.append(result)
                except Exception:
                    results.append(False)

            return all(results)

    def _commit_phase(self, services: List[object], tx_id: str, timeout: int) -> bool:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = []
            for service in services:
                futures.append(executor.submit(self._call_service_operation, 
                                             service.commit, tx_id, timeout))

            results = []
            for future in concurrent.futures.as_completed(futures):
                try:
                    future.result()
                    results.append(True)
                except Exception:
                    results.append(False)

            return all(results)

    def _rollback_phase(self, services: List[object], tx_id: str, timeout: int) -> bool:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = []
            for service in services:
                futures.append(executor.submit(self._call_service_operation, 
                                             service.rollback, tx_id, timeout))

            results = []
            for future in concurrent.futures.as_completed(futures):
                try:
                    future.result()
                    results.append(True)
                except Exception:
                    results.append(False)

            return all(results)

    def _call_service_operation(self, operation: Callable, tx_id: str, timeout: int):
        start_time = time.time()
        try:
            future = concurrent.futures.Future()
            
            def worker():
                try:
                    result = operation(tx_id)
                    future.set_result(result)
                except Exception as e:
                    future.set_exception(e)
            
            thread = threading.Thread(target=worker)
            thread.start()
            thread.join(timeout=timeout)
            
            if thread.is_alive():
                raise TimeoutError(f"Operation timed out after {timeout} seconds")
            
            if time.time() - start_time > timeout:
                raise TimeoutError(f"Operation timed out after {timeout} seconds")
                
            return future.result()
        except Exception as e:
            raise e

File: java/exercises/practice/unit_test_d76befa1_87e1_4f12_8a56_e65e92fd5f58/distributed_tx/.meta/src/main/java/TransactionCoordinator.java
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.*;

public class TransactionCoordinator {
    private static final Logger logger = LoggerFactory.getLogger(TransactionCoordinator.class);
    private static final int PREPARE_TIMEOUT_SECONDS = 5;
    private static final int MAX_RETRIES = 3;
    private static final long INITIAL_BACKOFF_MILLIS = 100;

    private final ExecutorService executor;

    public TransactionCoordinator() {
        this.executor = Executors.newCachedThreadPool();
    }

    public TransactionContext begin() {
        String transactionId = UUID.randomUUID().toString();
        logger.info("Starting transaction: {}", transactionId);
        return new TransactionContext(transactionId);
    }

    public void commit(List<Service> services, TransactionContext context) {
        logger.info("Starting prepare phase for transaction: {}", context.getTransactionId());
        List<Future<Boolean>> futures = new ArrayList<>();
        for (Service s : services) {
            Future<Boolean> future = executor.submit(() -> {
                try {
                    boolean result = s.prepare(context);
                    logger.info("Service {} prepare returned {}", s.getClass().getSimpleName(), result);
                    return result;
                } catch (Exception e) {
                    logger.error("Service {} prepare threw exception: {}", s.getClass().getSimpleName(), e.getMessage());
                    return false;
                }
            });
            futures.add(future);
        }

        boolean allPrepared = true;
        for (Future<Boolean> future : futures) {
            try {
                boolean result = future.get(PREPARE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
                if (!result) {
                    allPrepared = false;
                }
            } catch (TimeoutException te) {
                logger.error("Service prepare timed out after {} seconds", PREPARE_TIMEOUT_SECONDS);
                allPrepared = false;
            } catch (Exception e) {
                logger.error("Exception during prepare phase: {}", e.getMessage());
                allPrepared = false;
            }
        }

        if (!allPrepared) {
            logger.warn("Prepare phase failed for transaction: {}. Initiating rollback.", context.getTransactionId());
            rollback(services, context);
            throw new RuntimeException("Prepare phase failed. Transaction rolled back.");
        }

        logger.info("All services prepared successfully for transaction: {}. Starting commit phase.", context.getTransactionId());
        for (Service s : services) {
            executeWithRetry(() -> {
                s.commit(context);
                logger.info("Service {} commit successful.", s.getClass().getSimpleName());
                return null;
            });
        }
        logger.info("Transaction {} committed successfully.", context.getTransactionId());
    }

    public void rollback(List<Service> services, TransactionContext context) {
        logger.info("Starting rollback for transaction: {}", context.getTransactionId());
        for (Service s : services) {
            try {
                executeWithRetry(() -> {
                    s.rollback(context);
                    logger.info("Service {} rollback successful.", s.getClass().getSimpleName());
                    return null;
                });
            } catch (Exception e) {
                logger.error("Service {} rollback ultimately failed: {}", s.getClass().getSimpleName(), e.getMessage());
            }
        }
        logger.info("Rollback completed for transaction: {}", context.getTransactionId());
    }

    private <T> T executeWithRetry(Callable<T> callable) {
        long backoff = INITIAL_BACKOFF_MILLIS;
        for (int attempt = 0; attempt < MAX_RETRIES; attempt++) {
            try {
                return callable.call();
            } catch (Exception e) {
                logger.warn("Attempt {} failed with exception: {}. Retrying after {} ms", attempt + 1, e.getMessage(), backoff);
                try {
                    Thread.sleep(backoff);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException("Interrupted during retry backoff", ie);
                }
                backoff *= 2;
            }
        }
        try {
            return callable.call();
        } catch (Exception e) {
            logger.error("Operation failed after {} retries: {}", MAX_RETRIES, e.getMessage());
            throw new RuntimeException("Operation failed after retries", e);
        }
    }
}